File: src/main/scala/seclogin/math/RandomBigIntModQ.scala
package seclogin.math

import scala.collection.immutable.Set
import scala.collection.mutable
import scala.math.BigInt
import scala.util.Random
import java.security.SecureRandom
import java.math.BigInteger

/** Adapts a `Random` for use with `BigInt`s in a finite field `Mod` q.
  */
case class RandomBigIntModQ(r: Random = new SecureRandom) {

  /** A Java-friendly constructor.
    */
  def this(r: java.util.Random) =
    this(Random.javaRandomToRandom(r))

  /** @return a uniformly random element of ''ℤ,,q,,''
    */
  def nextBigIntModQ()(implicit q: Mod): BigInt = {
    bigIntModQStream().head
  }

  /** @return a uniformly random element of ''ℤ,,q,,'' not equal to the given value
    */
  def nextBigIntModQNotEqualTo(forbidden: BigInt)(implicit q: Mod): BigInt = {
    bigIntModQStream().filter(_ != forbidden).head
  }

  /** @return a uniformly random element of ''ℤ,,q,,'' not equal to the given value
    */
  def nextBigIntegerModQNotEqualTo(forbidden: BigInteger, q: BigInteger): BigInteger = {
    nextBigIntModQNotEqualTo(forbidden)(Mod(q)).bigInteger
  }

  private def bigIntModQStream()(implicit q: Mod): Stream[BigInt] = {
    val bits = q.bitLength
    Stream.continually(BigInt(numbits = bits, rnd = r)).filter(_ < q)
  }

  /** @return a uniformly random size-`n` subset of ''ℤ,,q,,''
    */
  def nextUniqueBigIntsModQ(n: Int)(implicit q: Mod): Set[BigInt] = {
    val set = mutable.Set[BigInt]()
    while (set.size < n) set += nextBigIntModQ()
    set.toSet
  }

}

File: src/main/java/seclogin/math/RandomBigIntModQ.java
package seclogin.math;

import java.math.BigInteger;
import java.util.HashSet;
import java.util.Random;
import java.util.Set;

/**
 * Adapts a `Random` for use with `BigInt`s in a finite field `Mod` q.
 */
public class RandomBigIntModQ {

    private final Random r;
    final Mod q;

    public RandomBigIntModQ(Random r, Mod q) {
        this.r = r;
        this.q = q;
    }

    /**
     * @return a uniformly random element of ''ℤ,,q,,''
     */
    public BigInteger nextBigIntModQ() {
        int bits = q.q.bitLength();
        while (true) {
            BigInteger x = new BigInteger(bits, r);
            if (x.compareTo(q.q) < 0) return x;
        }
    }

    /**
     * @return a uniformly random element of ''ℤ,,q,,'' not equal to the given value
     */
    public BigInteger nextBigIntegerModQNotEqualTo(BigInteger forbidden) {
        while (true) {
            BigInteger x = nextBigIntModQ();
            if (!x.equals(forbidden)) return x;
        }
    }

    /**
     * @return a uniformly random size-`n` subset of ''ℤ,,q,,''
     */
    Set<BigInteger> nextUniqueBigIntsModQ(int n) {
        Set<BigInteger> set = new HashSet<BigInteger>();
        while (set.size() < n) set.add(nextBigIntModQ());
        return set;
    }

}

